File: app/src/main/java/wangdaye/com/geometricweather/main/adapters/location/LocationModel.java
package wangdaye.com.geometricweather.main.adapters.location;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.List;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.basic.models.options.provider.WeatherSource;
import wangdaye.com.geometricweather.common.basic.models.options.unit.TemperatureUnit;
import wangdaye.com.geometricweather.common.basic.models.weather.Alert;
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode;
import wangdaye.com.geometricweather.common.utils.DisplayUtils;

public class LocationModel {

    public @NonNull Location location;

    public @Nullable WeatherCode weatherCode;
    public @Nullable String weatherText;
    public @NonNull WeatherSource weatherSource;
    public boolean currentPosition;
    public boolean residentPosition;

    public @NonNull String title;
    public @NonNull String body;

    public @Nullable String alerts;

    public boolean selected;

    public LocationModel(
            @NonNull Context context,
            @NonNull Location location,
            @NonNull TemperatureUnit unit,
            boolean selected
    ) {
        this.location = location;

        this.weatherCode = null;
        // TODO: Use current instead
        if (location.getWeather() != null && location.getWeather().getDailyForecast().size() > 0) {
            if (location.isDaylight()
                    && location.getWeather().getDailyForecast().get(0).getDay() != null
                    && location.getWeather().getDailyForecast().get(0).getDay().getWeatherCode() != null
            ) {
                this.weatherCode = location.getWeather().getDailyForecast().get(0).getDay().getWeatherCode();
                this.weatherText = location.getWeather().getDailyForecast().get(0).getDay().getWeatherText();
            }
            if (!location.isDaylight()
                    && location.getWeather().getDailyForecast().get(0).getNight() != null
                    && location.getWeather().getDailyForecast().get(0).getNight().getWeatherCode() != null
            ) {
                this.weatherCode = location.getWeather().getDailyForecast().get(0).getNight().getWeatherCode();
                this.weatherText = location.getWeather().getDailyForecast().get(0).getNight().getWeatherText();
            }
        }

        this.weatherSource = location.getWeatherSource();

        this.currentPosition = location.isCurrentPosition();
        this.residentPosition = location.isResidentPosition();

        this.title = location.isCurrentPosition()
                ? context.getString(R.string.current_location)
                : location.place();
        this.body = (location.isUsable()) ? location.administrationLevels() : context.getString(R.string.feedback_not_yet_location);

        if (location.getWeather() != null) {
            List<Alert> alertList = location.getWeather().getCurrentAlertList();
            if (alertList.size() > 0) {
                StringBuilder builder = new StringBuilder();
                for (int i = 0; i < alertList.size(); i++) {
                    builder.append(alertList.get(i).getDescription());
                    if (alertList.get(i).getStartDate() != null) {
                        String startDateDay = DisplayUtils.getFormattedDate(alertList.get(i).getStartDate(),
                                location.getTimeZone(),
                                context.getString(R.string.date_format_short));
                        builder.append(", ")
                            .append(startDateDay)
                            .append(", ")
                            .append(DisplayUtils.getFormattedDate(alertList.get(i).getStartDate(),
                                    location.getTimeZone(),
                                    (DisplayUtils.is12Hour(context)) ? "h:mm aa" : "HH:mm"));
                        if (alertList.get(i).getEndDate() != null) {
                            builder.append("-");
                            String endDateDay = DisplayUtils.getFormattedDate(alertList.get(i).getEndDate(),
                                    location.getTimeZone(),
                                    context.getString(R.string.date_format_short));
                            if (!startDateDay.equals(endDateDay)) {
                                builder.append(endDateDay)
                                        .append(", ");
                            }
                            builder.append(DisplayUtils.getFormattedDate(alertList.get(i).getEndDate(),
                                    location.getTimeZone(),
                                    (DisplayUtils.is12Hour(context)) ? "h:mm aa" : "HH:mm"));
                        }
                    }
                    if (i != alertList.size() - 1) {
                        builder.append("\n");
                    }
                }
                alerts = builder.toString();
            } else {
                alerts = null;
            }
        } else {
            alerts = null;
        }

        this.selected = selected;
    }

    public boolean areItemsTheSame(@NonNull LocationModel newItem) {
        return location.getFormattedId().equals(
                newItem.location.getFormattedId()
        );
    }

    public boolean areContentsTheSame(@NonNull LocationModel newItem) {
        return location.equals(newItem.location) && selected == newItem.selected;
    }
}

File: app/src/main/java/org/breezyweather/main/adapters/location/LocationModel.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.main.adapters.location

import android.content.Context
import org.breezyweather.R
import breezyweather.domain.location.model.Location
import org.breezyweather.common.basic.models.options.unit.TemperatureUnit
import breezyweather.domain.weather.model.WeatherCode
import org.breezyweather.common.extensions.getFormattedDate
import org.breezyweather.common.extensions.getFormattedTime
import org.breezyweather.common.extensions.is12Hour
import org.breezyweather.common.source.MainWeatherSource
import org.breezyweather.domain.location.model.getPlace

class LocationModel(
    context: Context,
    val location: Location,
    val mainWeatherSource: MainWeatherSource?,
    unit: TemperatureUnit, // TODO: Add back temperature
    var selected: Boolean
) {
    var weatherCode: WeatherCode? = null
    var weatherText: String? = null
    val currentPosition: Boolean = location.isCurrentPosition
    val title: String = location.getPlace(context, true)
    val body: String = if (location.isUsable) location.administrationLevels() else context.getString(R.string.location_current_not_found_yet)
    var alerts: String? = null

    init {
        location.weather?.let { weather ->
            weatherCode = weather.current?.weatherCode
            weatherText = weather.current?.weatherText
            if (weather.currentAlertList.isNotEmpty()) {
                val builder = StringBuilder()
                weather.currentAlertList.forEach { alert ->
                    if (builder.toString().isNotEmpty()) {
                        builder.append("\n")
                    }
                    builder.append(alert.description)
                    alert.startDate?.let { startDate ->
                        val startDateDay = startDate.getFormattedDate(
                            location.timeZone, context.getString(R.string.date_format_short)
                        )
                        builder.append(", ")
                            .append(startDateDay)
                            .append(", ")
                            .append(startDate.getFormattedTime(location.timeZone, context.is12Hour))
                        alert.endDate?.let { endDate ->
                            builder.append("-")
                            val endDateDay = endDate.getFormattedDate(
                                location.timeZone, context.getString(R.string.date_format_short)
                            )
                            if (startDateDay != endDateDay) {
                                builder.append(endDateDay)
                                    .append(", ")
                            }
                            builder.append(endDate.getFormattedTime(location.timeZone, context.is12Hour))
                        }
                    }
                }
                alerts = builder.toString()
            }
        }
    }

    fun areItemsTheSame(newItem: LocationModel) = location.formattedId == newItem.location.formattedId

    fun areContentsTheSame(newItem: LocationModel) = location == newItem.location && selected == newItem.selected
}
